File: JSstudyGroup/14502.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const data = [];

rl.on('line', (input) => {
  data.push(input);
}).on('close', () => {
  let [N, M] = data.shift().split(' ').map(Number);

  let lab = data.map((i) => i.split(' ').map(Number));

  // 0의 좌표 , 2의 좌표를 미리 받아 둠
  let zero = [];
  let two = [];
  for (let i = 0; i < lab.length; i++) {
    for (let j = 0; j < lab[i].length; j++) {
      if (lab[i][j] === 0) zero.push([i, j]);
      if (lab[i][j] === 2) two.push([i, j]);
    }
  }

  let [dx, dy] = [
    [-1, 1, 0, 0],
    [0, 0, -1, 1],
  ];

  // 0의 범위를 구하는 bfs
  const bfs = (lab) => {
    let queue = JSON.parse(JSON.stringify(two));

    while (queue.length) {
      let [x, y] = queue.shift();

      for (let i = 0; i < 4; i++) {
        let nx = x + dx[i];
        let ny = y + dy[i];
        if (nx >= 0 && nx < N && ny >= 0 && ny < M && lab[nx][ny] === 0) {
          lab[nx][ny] = 2;
          queue.push([nx, ny]);
        }
      }
    }

    const countZero = () => {
      let count = 0;
      for (let i = 0; i < lab.length; i++) {
        for (let j = 0; j < lab[i].length; j++) {
          if (lab[i][j] === 0) count += 1;
        }
      }
      return count;
    };

    return countZero();
  };


  // 1을 세울 수 있는 모든 경우의 수를 구하기 위한 백트래킹
  let result = [];
  const dfs = (index, depth) => {
    if (depth === 3) {

			// 반드시 깊은 복사로 넘겨줘야 한다
      let zeroZone = bfs(JSON.parse(JSON.stringify(lab)));
      result.push(zeroZone);
      return;
    }

    for (let i = index; i < zero.length; i++) {
      let [x, y] = zero[i];
      lab[x][y] = 1;
      dfs(i + 1, depth + 1);

      lab[x][y] = 0;
    }
  };
  
  dfs(0, 0);

  console.log(Math.max(...result));

  process.exit();
});

File: JSstudyGroup/14502.py
import copy
import collections

[N, M] = list(map(int, input().split(" ")))
board = [list(map(int, input().split(" "))) for _ in range(N)]
wallLocations = []
virus = []
zeroCount = []


[dx, dy] = [[-1, 1, 0, 0], [0, 0, -1, 1]]

for i in range(N):
    for j in range(M):
        value = board[i][j]
        if value == 2:
            virus.append([i, j])
        if board[i][j] == 0:
            zeroCount.append([i, j])


def bfs(board):
    copyBoard = copy.deepcopy(board)

    queue = collections.deque(virus)
    while len(queue):
        [x, y] = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx >= N or ny < 0 or ny >= M or copyBoard[nx][ny] != 0:
                continue
            copyBoard[nx][ny] = 2
            queue.append([nx, ny])

    count = 0
    for i in copyBoard:
        for j in i:
            if j == 0:
                count += 1

    return count


result = []


def dfs(board, x, y, currentWall):
    if len(currentWall) == 3:
        result.append(bfs(board))
        return

    for i in range(x, N):
        # 다음 행 부터는 0열부터 탐색해야 함
        if i != N:
            y = 0
        for j in range(y, M):
            value = board[i][j]
            if value == 0:
                board[i][j] = 1
                currentWall.append([i, j])
                # 다음 좌표부터 탐색해야 하므로
                dfs(board, i, j, currentWall)
                board[i][j] = 0
                currentWall.pop()


dfs(board, 0, 0, [])

print(max(result))
